### ISSUE TITLE
All join count methods should warn/fail for asymmetric weights

### ISSUE THREAD

**Issue by ljwolf** [pysal/esda#198]
[Over in the geographic data science book](https://github.com/darribas/geographic-data-science/commit/dcfdc5a195af355b3fd282c54b5333ca067889b8#diff-d10182504057eedfba88d256e0a171259e99356f5a4f92beb3632170369044e1R522), we just realized that join counts/local join count statistics are not well-defined in the case of asymmetric weights matrices (like k-nearest neighbors). 

Thus, we need to warn/error out if `w.asymmetries` is not an empty list.  I'm not sure how the statistics are defined in the directed case, and (regardless) we'd need to enhance their existing estimators for this case.

I think it's appropriate to raise a `NotImplementedError` immediately if the weights are asymmetric.

**Comment by sjsrey** [comment]
Is the issue the loss of the proportionality between the global JC and the sum of the local JCs with asymmetric weights, or something else?

**Comment by ljwolf** [comment]
Yes. Further, the join-count is only well-defined for symmetric weights, otherwise the Black-White join is not the same as a White-Black join. Our current testing procedure does not differentiate, and we test only one of these arbitrarily.
